src/engine/surface_collision.c

```diff
@@ -241,86 +203,61 @@ void resolve_and_return_wall_collisions(Vec3f pos, f32 offset, f32 radius, struc
  **************************************************/
 
 void add_ceil_margin(s32 *x, s32 *z, Vec3s target1, Vec3s target2, f32 margin) {
-	register f32 diff_x, diff_z, invDenom;
-	diff_x = target1[0] - *x + target2[0] - *x;
-	diff_z = target1[2] - *z + target2[2] - *z;
-	invDenom = margin / sqrtf(sqr(diff_x) + sqr(diff_z));
-	*x += diff_x * invDenom;
-	*z += diff_z * invDenom;
+    register f32 diff_x, diff_z, invDenom;
+    diff_x = target1[0] - *x + target2[0] - *x;
+    diff_z = target1[2] - *z + target2[2] - *z;
+    invDenom = margin / sqrtf(sqr(diff_x) + sqr(diff_z));
+    *x += diff_x * invDenom;
+    *z += diff_z * invDenom;
 }
 
 /**
  * Iterate through the list of ceilings and find the first ceiling over a given point.
  */
 static struct Surface *find_ceil_from_list(struct SurfaceNode *surfaceNode, s32 x, s32 y, s32 z, f32 *pheight) {
     const f32 margin = 1.5f;
-    register struct Surface *surf;
+    register struct Surface *surf, *ceil = NULL;
     Vec3i vx, vz;
     f32 height;
-    struct Surface *ceil = NULL;
+    s16 type;
     *pheight = CELL_HEIGHT_LIMIT;
     // Stay in this loop until out of ceilings.
     while (surfaceNode != NULL) {
         surf = surfaceNode->surface;
         surfaceNode = surfaceNode->next;
-        if (y > surf->upperY) {
+        type = surf->type;
+        // Determine if checking for the camera or not.
+        if (SURFACE_IS_NEW_WATER(type)) continue;
```

Comment:
i hate to say this, but this check if probably not necessary in walls/ceilings/floors since water goes to it's own list 😬

Reply:
wait.. maybe? 🤔 i'll have to get back on this one